src/lib/esnext.array.d.ts

```diff
@@ -0,0 +1,51 @@
+interface ReadonlyArray<T> {
+
+    /**
+     * Calls a defined callback function on each element of an array. Then, flattens the result into
+     * a new array.
+     * This is identical to a map followed by a flatten of depth 1.
+     *
+     * @param callback A function that accepts up to three arguments. The flatMap method calls the
+     * callback function one time for each element in the array.
+     * @param thisArg An object to which the this keyword can refer in the callback function. If
+     * thisArg is omitted, undefined is used as the this value.
+     */
+    flatMap<U, This = undefined> (
+        callback: (this: This, value: T, index: number, array: T[]) => U|U[],
+        thisArg?: This
+    ): U[]
+
+    /**
+     * Returns a new array with all sub-array elements concatenated into it recursively up to the
+     * specified depth. If no depth is provided, flatten method defaults to the depth of 1.
+     *
+     * @param depth The maximum recursion depth
+     */
+    flatten<U>(this: ReadonlyArray<U>, depth?: number): U[];
+  }
+
+interface Array<T> {
+
+    /**
+     * Calls a defined callback function on each element of an array. Then, flattens the result into
+     * a new array.
+     * This is identical to a map followed by a flatten of depth 1.
+     *
+     * @param callback A function that accepts up to three arguments. The flatMap method calls the
+     * callback function one time for each element in the array.
+     * @param thisArg An object to which the this keyword can refer in the callback function. If
+     * thisArg is omitted, undefined is used as the this value.
+     */
+    flatMap<U, This = undefined> (
+        callback: (this: This, value: T, index: number, array: T[]) => U|U[],
+        thisArg?: This
+    ): U[]

```

Comment:
This will, unfortunately, do the wrong thing if `U` is deduced to be an `Array`, but I'm not sure if there's a way of doing `<U extends anything but Array>`.

Reply:
Example:

```ts
const x = Array.from({ length: 2 }).flatMap((_, i) => (
  i === 0 ? [1] : [[2]]
))
// x is number[][] but should actually be Array<number|number[]>
```

Reply:
Indeed. However, I don't know either how we can achieve ` <U extends anything by Array>`.
Meanwhile, we can change the definition to 
`flatMap<U, This = undefined> (
        callback: (this: This, value: T, index: number, array: T[]) =>` __***U***__ `,thisArg?: This
    ): U[];
`
I will be available to continue working on this PR starting from December 15th.

Reply:
It's better if the callback's return is `U[]` instead of `U`, otherwise this looks no different, type-wise, from a regular `.map`.

Reply:
@Kovensky according to #20724, using conditional types it will be possible to define a type that extends anything but array.
With something like this I guess
``` typescript
type Diff<T, U> = T extends U ? never : T;
type NonArray<T> = Diff<T, Array>;

flatMap<U: NonArray, This = undefined> (
        callback: (this: This, value: T, index: number, array: T[]) => U|U[],
        thisArg?: This
): U[]
```

Reply:
@Kovensky for now, I will update the callback's return as `U[]` as you suggested